<p><code>Enumerable.Sum()</code> 总是在开启了 <code>checked</code> 溢出检查的情况下执行加法，因此如果值超过了 <code>MaxValue</code>就会抛出 <code>OverflowException</code>，即使手动启用了 <code>unchecked</code> 也是如此。总之在任何情况下，使用 <code>unchecked</code> 都是不熟悉 <code>Sum</code> 的工作原理的表现。</p>
<p>对整型的 <code>Sum</code> 使用 <code>unchecked</code> 时，此规则会引发问题。</p>
<h2>不合规的代码示例</h2>
<pre>
void Add(List&lt;int&gt; list)
{
  int d = unchecked(list.Sum());  // 不合规

  unchecked
  {
    int e = list.Sum();  // 不合规
  }
}
</pre>
<h2>合规的解决方案</h2>
<pre>
void Add(List&lt;int&gt; list)
{
  int d = list.Sum();

  try
  {
    int e = list.Sum();
  }
  catch (System.OverflowException e)
  {
    // 进行异常处理……
  }
}
</pre>
<h2>例外</h2>
<p>如果 <code>Sum()</code> 是在 <code>try-catch</code> 语句块中调用的，则不会报告问题。</p>
<pre>
void Add(List&lt;int&gt; list)
{
  unchecked
  {
    try
    {
      int e = list.Sum();
    }
    catch (System.OverflowException e)
    {
      // exception handling...
    }
  }
}
</pre>

